src/QueryObserver.ts

```diff
@@ -1,7 +1,101 @@
+import { Action, Query, QueryFunction } from "./Query"
 import { Subscribable } from "./subscribable"
+import { shallowEqualObjects } from "./utils"
 
-export class QueryObserver extends Subscribable {
-  onQueryUpdate(data: unknown): void {
-    console.log(data)
+export type QueryObserverListener<TData, TError> = (
+  result: QueryResult<TData, TError>
+) => void
+
+export interface QueryResult<TData, TError> {
+  data: TData | undefined
+  error: TError | null
+  status: "idle" | "error" | "loading" | "success"
+  isLoading: boolean
+  isError: boolean
+  isSuccess: boolean
+}
+
+export interface QueryObserverOptions<TData, TError> {
+  fetchOnMount?: boolean
+  listeners: QueryObserverListener<TData, TError>[]
+}
+
+const INITIAL_QUERY_OBSERVER_RESULT = {
+  data: undefined,
+  error: null,
+  status: "idle",
+  isLoading: false,
+  isError: false,
+  isSuccess: false,
+}
+
+export class QueryObserver<
+  TData = unknown,
+  TError = unknown
+> extends Subscribable<QueryObserverListener<TData, TError>> {
+  private currentResult: QueryResult<TData, TError>
```

Comment:
I don't wanna add bloat right away, but I was thinking down the line, if the lib really picks up in usage, it would be awesome to have the API in place to build some minimal dev tools / little GUI to time-travel through queries, so maybe down the line we might wanna have an array with past results.